### PR TITLE
Add a Array[String] argument of repo IDs to be enabled

### PR DESCRIPTION
**Pull request by pdemonaco**
#### Pull Request (PR) description

This request introduces three primary changes
1. Fact calculation is now constrained to the ['os']['name'] fact instead of the classic 'osfamily' fact. This ensures that the module doesn't waste time running on CentOS or other non RHEL RedHat derivatives.
2. The custom facts have been combined into a single 'rhsm' custom fact.
3. A new parameter named `enabled_repo_ids` has been added to the rhsm class to allow an array of strings.

#### This Pull Request (PR) fixes the following issues

* Directly addresses issue #68.
* Indirectly addresses issue #66 by allowing the desired repo to be specified.

### CODE REVIEW

**Comment by pdemonaco** [comment]
Huh. Is there a reason that the rubocop configuration for this module expects exactly the opposite behavior for trailing comments than that of the default in `pdk validate`?

File: .rspec [hunk 1]
```diff
@@ -1,2 +1,2 @@
---format documentation
 --color
+--format documentation
```
Comment (bastelfreak) [inline .rspec]: why did you change this?

File: lib/facter/rhsm_repos.rb [hunk 2]
```diff
@@ -1,11 +0,0 @@
-Facter.add('rhsm_repos') do
```
Comment (kallies) [inline lib/facter/rhsm_repos.rb]: Instead of removing the facts they might just output the chunk data from the all-in-one rhsm fact (backwards compatibility)

File: manifests/repo.pp [hunk 3]
```diff
@@ -12,9 +12,13 @@
 #   ::rhsm::repo { "rhel-${::operatingsystemmajrelease}-server-extras-rpms": }
 define rhsm::repo (
```
[the remark this answers is not in the document]
Comment (kallies) [reply manifests/repo.pp]: Thinking about it again, it would be nice to have a parameter for disabling repositories, too.

File: .rspec (hunk 1 above)
Comment (pdemonaco) [reply .rspec]: Shoot - I didn't mean to commit that. I converted the module to be PDK compatible for my own local testing during development and I accidentally added this file. PDK changed a ton of stuff and I'm not familiar enough with Travis or the testing configuration in use on this module so I didn't add the other changes. I'll revert this for now.

File: README.md [hunk 4]
```diff
@@ -51,6 +51,33 @@ You shouldn't specify the protocol, subscription-manager will use HTTP. For prox
...
+A string array of repo IDs can now be provided as an argument to the class definition. This list will be used to enable the target repos if that has not already occurred.
+
+The following example enables the server and optional RPMs:
+
+```puppet
+class { 'rhsm':
+  rh_user          => 'myuser',
+  rh_password      => 'mypassword',
+  enabled_repo_ids => [
+    'rhel-7-server-rpms',
```
[the remark this answers is not in the document]
Comment (pdemonaco) [reply README.md]: Unfortunately no, I don't have Satellite in my environment so I haven't tested anything like that.

**Comment by pdemonaco** [comment]
@kallies should I add a version bump in this pull request? I'm assuming by major you mean 4.x.x?

**Comment by kallies** [comment]
> @kallies should I add a version bump in this pull request? I'm assuming by major you mean 4.x.x?

Yes, 4.0.0 would be nice. If there are no braking changes (no fact removal etc.) 3.2.0 is fine for me, too.

File: lib/facter/rhsm.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
+Facter.add(:rhsm, type: :aggregate) do
+  confine :os do |os|
+    os['name'] == 'RedHat'
```
Comment (kallies) [inline lib/facter/rhsm.rb]: What about CentOS? subscription-manager is also available there.

Comment (pdemonaco) [reply lib/facter/rhsm.rb]: Hmm, what would the point be? Can CentOS use subscription-manager to work with custom satellite repos?

I was trying to exclude it to prevent my CentOS machines from failing because they don't have subscription-manger installed by default.

Comment (pdemonaco) [reply lib/facter/rhsm.rb]: No big deal either way - I've updated the calculation of the facts so they won't break on machines which don't have subscription-manager installed.

File: manifests/init.pp [hunk 6]
```diff
@@ -72,8 +73,7 @@
   $manage_repos          = 1,
   $full_refresh_on_yum   = 0,
   $package_ensure        = 'latest',
-  $repo_extras           = false,
-  $repo_optional         = false
+  Optional[Array[String]] $enabled_repo_ids = [],
```
Comment (bastelfreak) [inline manifests/init.pp]: Can you please use `Array[String[1]]` instead? The default for arrays should not be optional, but empty.  We documented that in [our review guidelines](https://voxpupuli.org/docs/#reviewing-a-module-pr)

File: manifests/init.pp [hunk 7]
```diff
@@ -157,16 +157,11 @@
     require => Exec['RHSM-register'],
   }
 
-  if $repo_extras {
-    ::rhsm::repo { "rhel-${::operatingsystemmajrelease}-server-extras-rpms":
-      require => Exec['RHSM-register'],
+  unless(empty($enabled_repo_ids)) {
+    $enabled_repo_ids.each | $repo_id | {
+      ::rhsm::repo { $repo_id:
```
Comment (bastelfreak) [inline manifests/init.pp]: please do not add the leading `::`, they are not needed anymore since we do not support Puppet3.

File: spec/default_facts.yml [hunk 8]
```diff
@@ -1,12 +1,6 @@
-# This file is managed via modulesync
-# https://github.com/voxpupuli/modulesync
-# https://github.com/voxpupuli/modulesync_config
+# Use default_module_facts.yml for module specific facts.
```
Comment (bastelfreak) [inline spec/default_facts.yml]: Please undo the changes here, otherwise our https://github.com/voxpupuli/modulesync_config will overwrite it on the next run and creates an unneeded diff.

File: spec/defines/repo_spec.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+          'rhsm' => {
+            'enabled_repo_ids' => []
+          }
+        )
+      end
+
+      context 'enabling server rpms' do
+        let(:title) { 'rhel-7-server-rpms' }
+
+        it { is_expected.to compile }
```
Comment (bastelfreak) [inline spec/defines/repo_spec.rb]: can you update it to `compile.with_all_deps` ?

File: spec/defines/repo_spec.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+          facts.merge(
+            'rhsm' => {
+              'enabled_repo_ids' => [
+                'rhel-7-optional-rpms'
+              ]
+            }
+          )
+        end
+
+        it { is_expected.to compile }
```
Comment (bastelfreak) [inline spec/defines/repo_spec.rb]: same here

**Comment by bastelfreak** [comment]
Because of the changed fact this is a breaking change and requires a new major release. You do not need to bump any version here, this will be triggered through the label I applied, during the next release. However, we need to ensure that an upgrade works smoothly for the users. I'm currently not sure how we should properly test it.

**Comment by pdemonaco** [comment]
@bastelfreak is this still backwards incompatible with the re-introduced 'rhsm_' facts? I also converted the confines statement to target the RedHat family again but caused it to skip execution for the list of enabled repos if the subscription-manager binary is missing.

File: README.md [hunk 11]
```diff
@@ -74,3 +111,7 @@ class { 'rhsm':
 ## Limitations
 
 Well, only RedHat is supported :)
+
+## Development
+
+Some general guidelines on PR structure can be found [here](https://voxpupuli.org/docs/#reviewing-a-module-pr).
```
Comment (bastelfreak) [inline README.md]: neat, we should add that to all modules.

File: manifests/init.pp [hunk 12]
```diff
@@ -72,8 +73,7 @@
   $manage_repos          = 1,
   $full_refresh_on_yum   = 0,
   $package_ensure        = 'latest',
-  $repo_extras           = false,
-  $repo_optional         = false
+  Array[String[1]] $enabled_repo_ids = [],
```
Comment (bastelfreak) [inline manifests/init.pp]: can you please align the `=`?

**Comment by bastelfreak** [comment]
@pdemonaco you're right, I think this is fine now. Please align the `=` and afterwards we can merge this.

**Comment by pdemonaco** [comment]
@bastelfreak will this generate a new version on the puppet forge automatically?

**Comment by bastelfreak** [comment]
No, we will have to do a new release, maybe in the next few days.

**Comment by ubellavance** [comment]
That looks like an awesome PR! I can't wait to try it (when I find time to do so). Thanks!
